shared/src/main/scala/mlscript/ucs/stages/Normalization.scala

```diff
@@ -0,0 +1,346 @@
+
+package mlscript.ucs.stages
+
+import mlscript.{App, CaseOf, Fld, FldFlags, Let, Loc, Sel, Term, Tup, Var, StrLit}
+import mlscript.{CaseBranches, Case, Wildcard, NoCases}
+import mlscript.Message, Message.MessageContext
+import mlscript.utils._, shorthands._
+import mlscript.ucs, mlscript.pretyper
+import ucs.{Desugarer, Lines, LinesOps, VariableGenerator}
+import ucs.context.{Context, Scrutinee}
+import ucs.display.{showNormalizedTerm, showSplit}
+import ucs.syntax.core.{Pattern, Branch, Split}
+import pretyper.Scope
+import pretyper.symbol._
+import pretyper.{Diagnosable, Traceable}
+
+trait Normalization { self: Desugarer with Traceable =>
+  import Normalization._
+
+  // TODO: We might not need the case where `deep` is `false`.
+  private def fillImpl(these: Split, those: Split, deep: Bool)(implicit
+      scope: Scope,
+      context: Context,
+      generatedVars: Set[Var],
```

Comment:
Avoid using implicit set parameters to prevent the following implicit conversion problem:
```scala
def foo(implicit v: Set[String]) = {
  if ("abc") 42 else 0
}
```